targets/psa_switch/psa_switch.cpp

```diff
@@ -313,9 +314,19 @@ PsaSwitch::ingress_thread() {
 
     phv = packet->get_phv();
 
+    // PSA wants us to resubmit the instance of the packet BEFORE ingress
+    // processing, so make a copy before ingress touches the packet
+    std::unique_ptr<Packet> packet_copy = packet->clone_no_phv_ptr();
+    PHV *phv_copy = packet_copy->get_phv();
+    phv_copy->reset_metadata();
```

Comment:
I do not know the memory alloc/dealloc or performance implications of doing resubmit this way, vs. the way it is done in simple_switch, which _also_ resubmits the packet as it was before ingress processing began.  @antoninbas can comment  more intelligently than I can there.

Reply:
@jafingerhut ours is the "wasteful/naive" method.

I think @antoninbas used a buffer mechanism for the packet to store a copy of itself (the entire unprased packet) before v1model parsing? But maybe I'm wrong haha since there's some syntax in packet_buffer that I haven't seen before, his comments would be more valuable here. 

Basing my understanding off of the following:

https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch/simple_switch.cpp#L497

https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/packet.h#L181

https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/packet_buffer.h#L103

Reply:
`buffer_state_t` doesn't copy anything, it just stores an offset. You currently make a copy for all ingress packets, even the ones that won't be resubmitted. It's not more complicated to use `buffer_state_t`, since that code already exists in simple_switch. Ideally, you may be able to avoid any kind of copying for resubmit, like @derekso1 did for recirculate. But if there is no way to do that, making a copy *only* for packets which are resubmitted is fine.

Reply:
We have to make copies for packets ahead of time unless we know which packets are resubmit ahead of ingress MAU. Currently this is the only solution we could come up with after discussing with Andy. Please let us know if any alternative ideas come up :)

Reply:
I already pointed you to the simple_switch code: https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch/simple_switch.cpp#L497, which doesn't do a copy.

Reply:
Kevin, I believe that the idea of the simple_switch implementation is that even though it does not know before executing ingress code whether the packet will be resubmitted or not, simple_switch does not modify the original packet until after ingress is complete (and perhaps not in any case does it ever modify it?  Not sure about that).  Thus resubmit with the original packet works.  (Do not trust me blindly on that explanation, as I may be slightly off somehow).

Reply:
Hi so I've been working with Kevin on this PR. We have a test that passes only if the original packet has been modified directly after ingress MAU and infinite loops otherwise.

With the simple switch code in place of what we have, we are passing that test. I'll go back and double check but we reiterated over this twice and observed the same behavior which lead us to the comment from above. Would it perhaps be best to get on a call briefly to discuss?

Reply:
A Packet object is essentially the packet buffer + PHV (set of extracted fields and metadata). The important thing is that for the resubmitted packet, the packet buffer be the same as for the original packet. The PHV will be "initialized" by the parser + by the reset metadata operation (which is really optional for PSA). The parser implementation modifies the packet buffer (the payload pointer is moved every time an header is extracted), which is why the packet buffer state needs to be saved (https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch/simple_switch.cpp#L497) before parsing and restored (https://github.com/p4lang/behavioral-model/blob/master/targets/simple_switch/simple_switch.cpp#L582) after the MAU pipeline. This basically requires 2 pointer copies instead of a much more expensive packet copy.

Note that I'm assuming the original packet does not go through the ingress deparser when we resubmit it. I think the PSA packet paths diagram (https://p4.org/p4-spec/docs/PSA-v1.1.0.html#packet-paths) is misleading. Since the resubmitted packet uses the original headers, the deparser is not involved.

Reply:
resolved now

Reply:
https://github.com/p4lang/behavioral-model/pull/773/files#diff-5ae8a8504eb56bc8167d4b69469ed7e8R339